### ISSUE TITLE
Connecting to peers errors out

### ISSUE THREAD

**Issue by fishseabowl** [pyrsia/pyrsia#865]
The user starts node A, and starts node B to connect to node A, then runs "pyrsia status", but cannot find any connected nodes.
```
./target/debug/pyrsia status
Connected Peers Count:       0
```

The test steps and logs are as follows:
Test case 1: Start node A with pyrsia debug and info log in terminal 1, start node B with pyrsia debug and info log in terminal 2, and check pyrsia status in terminal 3
1. Start node A in terminal 1
```
DEV_MODE=on RUST_LOG="pyrsia=debug,info"  ./target/debug/pyrsia_node -H 0.0.0.0 -p 7888 -L /ip4/0.0.0.0/tcp/44001
 DEBUG pyrsia_node > Parse CLI arguments
 DEBUG pyrsia_node > Create p2p components
 DEBUG pyrsia_node > Create artifact service
 DEBUG pyrsia_node > Create blockchain components
 DEBUG pyrsia_node > Start p2p event loop
 DEBUG pyrsia_node > Setup HTTP server
 DEBUG pyrsia_node > Pyrsia Docker Node will bind to host = 0.0.0.0, port = 7888
 DEBUG pyrsia_node > Setup HTTP routing
 DEBUG pyrsia_node > Setup HTTP server
 INFO  pyrsia_node > Pyrsia Docker Node will start running on 0.0.0.0:7888
 DEBUG pyrsia_node > Start p2p components
 DEBUG pyrsia::network::client > p2p::Client::listen "/ip4/0.0.0.0/tcp/44001"
 DEBUG pyrsia_node             > Listen for p2p events
 INFO  pyrsia::network::event_loop > Local node is listening on "/ip4/100.64.0.1/tcp/44001/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g"
 INFO  pyrsia::network::event_loop > Local node is listening on "/ip4/127.0.0.1/tcp/44001/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g"
 INFO  pyrsia::network::event_loop > Local node is listening on "/ip4/10.6.104.146/tcp/44001/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g"
 INFO  pyrsia_registry             > 127.0.0.1:50023 "GET /status HTTP/1.1" 200 "-" "-" 5.203791ms
```

2. Start node B in terminal 2 on another directory
 ```
DEV_MODE=on RUST_LOG="pyrsia=debug,info"  ./target/debug/pyrsia_node -H 0.0.0.0 -p 7889 -L /ip4/0.0.0.0/tcp/44002 --peer /ip4/127.0.0.1/tcp/44001/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g
  DEBUG pyrsia_node > Parse CLI arguments
  DEBUG pyrsia_node > Create p2p components
  DEBUG pyrsia_node > Create artifact service
  DEBUG pyrsia_node > Create blockchain components
  DEBUG pyrsia_node > Start p2p event loop
  DEBUG pyrsia_node > Setup HTTP server
  DEBUG pyrsia_node > Pyrsia Docker Node will bind to host = 0.0.0.0, port = 7889
  DEBUG pyrsia_node > Setup HTTP routing
  DEBUG pyrsia_node > Setup HTTP server
  INFO  pyrsia_node > Pyrsia Docker Node will start running on 0.0.0.0:7889
  DEBUG pyrsia_node > Start p2p components
  DEBUG pyrsia::network::client > p2p::Client::listen "/ip4/0.0.0.0/tcp/44002"
  DEBUG pyrsia::network::client > p2p::Client::dial "/ip4/127.0.0.1/tcp/44001/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g"
  INFO  pyrsia::network::event_loop > Local node is listening on "/ip4/100.64.0.1/tcp/44002/p2p/12D3KooWG8WRnhTnVKHpSDKtz9LkCPQ3F6TJTYmRFZYAoy3JkEhm"
  INFO  pyrsia::network::event_loop > Local node is listening on "/ip4/10.6.104.146/tcp/44002/p2p/12D3KooWG8WRnhTnVKHpSDKtz9LkCPQ3F6TJTYmRFZYAoy3JkEhm"
  INFO  pyrsia::network::event_loop > Local node is listening on "/ip4/127.0.0.1/tcp/44002/p2p/12D3KooWG8WRnhTnVKHpSDKtz9LkCPQ3F6TJTYmRFZYAoy3JkEhm"
  DEBUG pyrsia_node                 > Listen for p2p events
```
3. Check Pyrsia status in terminal 3
```
./target/debug/pyrsia status
Connected Peers Count:       0
```

Test case 2: Start node A with libp2p debug and info log in terminal 1, start node B with libp2 debug and info log in terminal 2, and check pyrsia status in terminal 3
1. Start node A in terminal 1
```
DEV_MODE=on RUST_LOG="libp2p=debug,info"  ./target/debug/pyrsia_node -H 0.0.0.0 -p 7888 -L /ip4/0.0.0.0/tcp/44001
 INFO  pyrsia_node > Pyrsia Docker Node will start running on 0.0.0.0:7888
 DEBUG libp2p_tcp  > listening on 0.0.0.0:44001
 DEBUG libp2p_tcp  > New listen address: /ip4/127.0.0.1/tcp/44001
 DEBUG libp2p_swarm > Listener ListenerId(18013658767067450256); New address: "/ip4/127.0.0.1/tcp/44001"
 INFO  pyrsia::network::event_loop > Local node is listening on "/ip4/127.0.0.1/tcp/44001/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g"
 DEBUG libp2p_tcp                  > New listen address: /ip4/10.6.104.146/tcp/44001
 DEBUG libp2p_swarm                > Listener ListenerId(18013658767067450256); New address: "/ip4/10.6.104.146/tcp/44001"
 INFO  pyrsia::network::event_loop > Local node is listening on "/ip4/10.6.104.146/tcp/44001/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g"
 DEBUG libp2p_tcp                  > New listen address: /ip4/100.64.0.1/tcp/44001
 DEBUG libp2p_swarm                > Listener ListenerId(18013658767067450256); New address: "/ip4/100.64.0.1/tcp/44001"
 INFO  pyrsia::network::event_loop > Local node is listening on "/ip4/100.64.0.1/tcp/44001/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g"
 DEBUG libp2p_tcp                  > Incoming connection from /ip4/127.0.0.1/tcp/50205 at /ip4/127.0.0.1/tcp/44001
 DEBUG libp2p_core::upgrade::apply > Successfully applied negotiated protocol
 DEBUG libp2p_core::upgrade::apply > Successfully applied negotiated protocol
 DEBUG libp2p_swarm                > Connection established: PeerId("12D3KooWG8WRnhTnVKHpSDKtz9LkCPQ3F6TJTYmRFZYAoy3JkEhm") Listener { local_addr: "/ip4/127.0.0.1/tcp/44001", send_back_addr: "/ip4/127.0.0.1/tcp/50205" }; Total (peer): 1. Total non-banned (peer): 1
 INFO  pyrsia_registry             > 127.0.0.1:50211 "GET /status HTTP/1.1" 200 "-" "-" 1.377496ms
 DEBUG libp2p_swarm                > Connection closed with error IO(Custom { kind: Other, error: A(YamuxError(Closed)) }): Connected { endpoint: Listener { local_addr: "/ip4/127.0.0.1/tcp/44001", send_back_addr: "/ip4/127.0.0.1/tcp/50205" }, peer_id: PeerId("12D3KooWG8WRnhTnVKHpSDKtz9LkCPQ3F6TJTYmRFZYAoy3JkEhm") }; Total (peer): 0.
```

2. Start node B in terminal 2 on another directory
 ```
DEV_MODE=on RUST_LOG="libp2p=debug,info"  ./target/debug/pyrsia_node -H 0.0.0.0 -p 7889 -L /ip4/0.0.0.0/tcp/44002 --peer /ip4/127.0.0.1/tcp/44001/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g
 INFO  pyrsia_node > Pyrsia Docker Node will start running on 0.0.0.0:7889
 DEBUG libp2p_tcp  > listening on 0.0.0.0:44002
 DEBUG libp2p_tcp  > New listen address: /ip4/100.64.0.1/tcp/44002
 DEBUG libp2p_swarm > Listener ListenerId(149982185294582183); New address: "/ip4/100.64.0.1/tcp/44002"
 INFO  pyrsia::network::event_loop > Local node is listening on "/ip4/100.64.0.1/tcp/44002/p2p/12D3KooWG8WRnhTnVKHpSDKtz9LkCPQ3F6TJTYmRFZYAoy3JkEhm"
 DEBUG libp2p_tcp                  > New listen address: /ip4/10.6.104.146/tcp/44002
 DEBUG libp2p_swarm                > Listener ListenerId(149982185294582183); New address: "/ip4/10.6.104.146/tcp/44002"
 INFO  pyrsia::network::event_loop > Local node is listening on "/ip4/10.6.104.146/tcp/44002/p2p/12D3KooWG8WRnhTnVKHpSDKtz9LkCPQ3F6TJTYmRFZYAoy3JkEhm"
 DEBUG libp2p_tcp                  > New listen address: /ip4/127.0.0.1/tcp/44002
 DEBUG libp2p_swarm                > Listener ListenerId(149982185294582183); New address: "/ip4/127.0.0.1/tcp/44002"
 INFO  pyrsia::network::event_loop > Local node is listening on "/ip4/127.0.0.1/tcp/44002/p2p/12D3KooWG8WRnhTnVKHpSDKtz9LkCPQ3F6TJTYmRFZYAoy3JkEhm"
 DEBUG libp2p_dns                  > Dialing /ip4/127.0.0.1/tcp/44001/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g
 DEBUG libp2p_tcp                  > dialing 127.0.0.1:44001
 DEBUG libp2p_core::upgrade::apply > Successfully applied negotiated protocol
 DEBUG libp2p_core::upgrade::apply > Successfully applied negotiated protocol
 DEBUG libp2p_swarm                > Connection established: PeerId("12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g") Dialer { address: "/ip4/127.0.0.1/tcp/44001/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g", role_override: Dialer }; Total (peer): 1. Total non-banned (peer): 1
 DEBUG libp2p_swarm                > Connection closed with error KeepAliveTimeout: Connected { endpoint: Dialer { address: "/ip4/127.0.0.1/tcp/44001/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g/p2p/12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g", role_override: Dialer }, peer_id: PeerId("12D3KooWLivw3St3KmaEEBUGrNsEKAFbCrn3bQQEZvxSGPrJX39g") }; Total (peer): 0.
```
3. Check Pyrsia status in terminal 3
```
./target/debug/pyrsia status
Connected Peers Count:       0
```

**Comment by tiainen** [comment]
That's perfectly expected behavior. Kademlia is configured by default to close connections after 10 seconds of inactivity (see https://docs.rs/libp2p-kad/latest/src/libp2p_kad/behaviour.rs.html#209). There's no reason to keep the connection alive if there is no activity happening. The connection will be automatically re-established when needed. Have a look at the following logs:

These are the logs when starting the second pyrsia node, that dials a running boot node:
```
[2022-07-19T06:45:28Z DEBUG pyrsia_node] Start p2p components
[2022-07-19T06:45:28Z DEBUG pyrsia::network::client] p2p::Client::listen "/ip4/127.0.0.1/tcp/44001"
[2022-07-19T06:45:28Z DEBUG libp2p_tcp] listening on 127.0.0.1:44001
[2022-07-19T06:45:28Z DEBUG libp2p_swarm] Listener ListenerId(1); New address: "/ip4/127.0.0.1/tcp/44001"
[2022-07-19T06:45:28Z DEBUG pyrsia::network::client] p2p::Client::dial "/ip4/127.0.0.1/tcp/44000/p2p/12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi"
[2022-07-19T06:45:28Z INFO  pyrsia::network::event_loop] Local node is listening on "/ip4/127.0.0.1/tcp/44001/p2p/12D3KooWN8QuDfNsA39azto1PgVoNk7qb1raHSLoXhpeoRtNUmyD"
[2022-07-19T06:45:28Z DEBUG libp2p_dns] Dialing /ip4/127.0.0.1/tcp/44000/p2p/12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi/p2p/12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi
[2022-07-19T06:45:28Z DEBUG libp2p_tcp] dialing 127.0.0.1:44000
[2022-07-19T06:45:28Z DEBUG libp2p_swarm] Connection established: PeerId("12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi") Dialer { address: "/ip4/127.0.0.1/tcp/44000/p2p/12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi/p2p/12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi", role_override: Dialer }; Total (peer): 1. Total non-banned (peer): 1
[2022-07-19T06:45:28Z DEBUG pyrsia_node] Listen for p2p events
```

After some time, the connection gets closed:
```
[2022-07-19T06:45:48Z DEBUG libp2p_swarm] Connection closed with error KeepAliveTimeout: Connected { endpoint: Dialer { address: "/ip4/127.0.0.1/tcp/44000/p2p/12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi/p2p/12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi", role_override: Dialer }, peer_id: PeerId("12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi") }; Total (peer): 0.
```

Then I request an artifact via the p2p network and the connection is re-established:
```
[2022-07-19T06:46:23Z INFO  pyrsia::artifact_service::storage] An artifact is being pulled from the artifact manager e96e8f17-6894-426c-907c-61dd70c70160
[2022-07-19T06:46:23Z DEBUG pyrsia::network::client] p2p::Client::list_providers "e96e8f17-6894-426c-907c-61dd70c70160"
[2022-07-19T06:46:23Z DEBUG pyrsia::network::event_loop] Local Peer 12D3KooWN8QuDfNsA39azto1PgVoNk7qb1raHSLoXhpeoRtNUmyD is dialing Peer 12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi...
[2022-07-19T06:46:23Z DEBUG libp2p_dns] Dialing /ip4/127.0.0.1/tcp/44000/p2p/12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi
[2022-07-19T06:46:23Z DEBUG libp2p_tcp] dialing 127.0.0.1:44000
[2022-07-19T06:46:23Z DEBUG libp2p_swarm] Connection established: PeerId("12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi") Dialer { address: "/ip4/127.0.0.1/tcp/44000/p2p/12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi", role_override: Dialer }; Total (peer): 1. Total non-banned (peer): 1
[2022-07-19T06:46:23Z DEBUG pyrsia::network::client] p2p::Client::get_idle_peer() entered with 2 peers
[2022-07-19T06:46:23Z DEBUG pyrsia::network::client] p2p::Client::get_idle_peer() Unable to get peer metric for peer 12D3KooWN8QuDfNsA39azto1PgVoNk7qb1raHSLoXhpeoRtNUmyD error Failed to dial the requested peer
[2022-07-19T06:46:23Z DEBUG pyrsia::network::idle_metric_protocol] p2p::idle_metric_protocol::write_request writing a request to peer for and idle metric
[2022-07-19T06:46:23Z DEBUG pyrsia::network::idle_metric_protocol] p2p::idle_metric_protocol::read_response Reading response to idle metric request with value =[228, 165, 155, 64, 41, 56, 143, 65]
[2022-07-19T06:46:23Z DEBUG pyrsia::network::client] p2p::Client::get_idle_peer() Pushing idle peer with value 65471784.076000005
[2022-07-19T06:46:23Z DEBUG pyrsia::network::client] p2p::Client::request_artifact PeerId("12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi"): "e96e8f17-6894-426c-907c-61dd70c70160"
[2022-07-19T06:46:23Z DEBUG pyrsia::network::artifact_protocol] Write ArtifactRequest: "e96e8f17-6894-426c-907c-61dd70c70160"
[2022-07-19T06:46:23Z INFO  pyrsia::artifact_service::service] put_artifact with id: e96e8f17-6894-426c-907c-61dd70c70160
[2022-07-19T06:46:23Z INFO  pyrsia::artifact_service::storage] An artifact is being pushed to the artifact manager e96e8f17-6894-426c-907c-61dd70c70160
[2022-07-19T06:46:23Z INFO  pyrsia::artifact_service::storage] An artifact is being pulled from the artifact manager e96e8f17-6894-426c-907c-61dd70c70160
[2022-07-19T06:46:23Z INFO  pyrsia_registry] 127.0.0.1:42876 "GET /maven2/commons-codec/commons-codec/1.15/commons-codec-1.15.pom HTTP/1.1" 200 "-" "curl/7.68.0" 212.864216ms
```

And after that the connection gets closed again, as is to be expected:
```
[2022-07-19T06:46:43Z DEBUG libp2p_swarm] Connection closed with error KeepAliveTimeout: Connected { endpoint: Dialer { address: "/ip4/127.0.0.1/tcp/44000/p2p/12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi", role_override: Dialer }, peer_id: PeerId("12D3KooWP7fGd6Ujk1XpCVYezhSL9j9zCJqN8qDEQQDXMfVDS9gi") }; Total (peer): 0.
```

**Comment by fishseabowl** [comment]
@tiainen  Thanks for your comments. But from the user side, it may not be friendly. The user starts node A, starts node B to connect to node A, then runs "pyrsia status", but cannot find any connected nodes.

**Comment by tiainen** [comment]
That's a different issue than what is described here. The reason for that is explained here: https://docs.rs/libp2p-kad/latest/libp2p_kad/index.html#important-discrepancies

**Comment by fishseabowl** [comment]
@tiainen Thanks for your comments. I have changed the title and description of this issue to be more clear.

**Comment by betarelease** [comment]
Updated the title so that it is a bug and not just an explore.

**Comment by tiainen** [comment]
This issue is not a real issue. I've closed this in favor of #886 which explains better what we have at the moment and what is needed.
